File: 백준/Gold/15644. 구슬 탈출 3/구슬 탈출 3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    private static int N, M, oriRedY, oriRedX, oriBlueY, oriBlueX, answer;
    private static char[][] lst;
    private static int[][][][] nextPosition;
    private static int[] nextRedPosition, nextBluePosition;
    private static int[] answerLst;

    public static void main(String[] args) throws IOException {

        int[][] dxy = new int[][] {{-1, 0, 0}, {0, 1, 1}, {1, 0, 2}, {0, -1, 3}};

        st = new StringTokenizer(br.readLine());
        N  = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        lst = new char[N][M];
        nextPosition = new int[N][M][4][2];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            String s = st.nextToken();
            for (int j = 0; j < M; j++) {
                lst[i][j] = s.charAt(j);
            }
        } // 입력 완료

        for (int i = 1; i < N - 1; i++) {
            for (int j = 1; j < M - 1; j++) {
                if (lst[i][j] == 'R') {
                    oriRedY = i;
                    oriRedX = j;
                } else if (lst[i][j] == 'B') {
                    oriBlueY = i;
                    oriBlueX = j;
                }
                for (int[] nxt: dxy) {
                    for (int k = 0; k < Math.max(N, M); k++) {
                        int ny = i + nxt[0] * k;
                        int nx = j + nxt[1] * k;
                        if (ny < 0 || ny >= N || nx < 0 || nx >= M) break;
                        if (lst[ny][nx] == 'O') {
                            nextPosition[i][j][nxt[2]][0] = 20;
                            nextPosition[i][j][nxt[2]][1] = 20;
                            break;
                        } else if (lst[ny][nx] == 'R' || lst[ny][nx] == 'B' || lst[ny][nx] == '.') {
                            nextPosition[i][j][nxt[2]][0] = ny;
                            nextPosition[i][j][nxt[2]][1] = nx;
                        } else {
                            break;
                        }
                    }
                }
            }
        } // end of for i

        answer = 20;
        answerLst = new int[10];
        recur(0, new int[] {oriRedY, oriRedX}, new int[] {oriBlueY, oriBlueX}, new int[10]);

        if (answer != 20) {
            System.out.println(answer);
            for (int i = 0; i < answer; i++) {
                switch (answerLst[i]) {
                    case 0:
                        System.out.print("U");
                        break;
                    case 1:
                        System.out.print("R");
                        break;
                    case 2:
                        System.out.print("D");
                        break;
                    case 3:
                        System.out.print("L");
                        break;
                }
            }
        } else {
            System.out.println(-1);
        }
    } // end of main

    private static void recur(int cur, int[] redPosition, int[] bluePosition, int[] history) {
        if (cur == 10 || cur + 1 >= answer) return;
        nextRedPosition = new int[2];
        nextBluePosition = new int[2];
        for (int i = 0; i < 4; i++) {
            int[][] res = proc(i, redPosition, bluePosition);
            history[cur] = i;
            if (res[0][0] == 1) {
                if (answer > cur + 1) {
                    answer = cur + 1;
                    for (int j = 0; j < answer; j++) {
                        answerLst[j] = history[j];
                    }
                }
                return;
            } else if (res[0][0] == 2) {
                continue;
            } else {
                nextRedPosition = res[1];
                nextBluePosition = res[2];
            }
            recur(cur + 1, nextRedPosition, nextBluePosition, history);
        }
    } // end of recur

    private static int[][] proc(int d, int[] redPosition, int[] bluePosition) {
        int nextRedY = nextPosition[redPosition[0]][redPosition[1]][d][0];
        int nextRedX = nextPosition[redPosition[0]][redPosition[1]][d][1];
        int nextBlueY = nextPosition[bluePosition[0]][bluePosition[1]][d][0];
        int nextBlueX = nextPosition[bluePosition[0]][bluePosition[1]][d][1];
        if (nextBlueY == 20) {
            return new int[][] {{2}};
        } else if (nextRedY == 20) {
            return new int[][] {{1}};
        }
        if (nextRedY == nextBlueY && nextRedX == nextBlueX) {
            if (d == 0) {
                if (redPosition[0] > bluePosition[0])
                    nextRedY += 1;
                else
                    nextBlueY += 1;
            } else if (d == 1) {
                if (redPosition[1] > bluePosition[1])
                    nextBlueX -= 1;
                else
                    nextRedX -= 1;
            } else if (d == 2) {
                if (redPosition[0] > bluePosition[0])
                    nextBlueY -= 1;
                else
                    nextRedY -= 1;
            } else {
                if (redPosition[1] > bluePosition[1])
                    nextRedX += 1;
                else
                    nextBlueX += 1;
            }
        }
        return new int[][] {{0}, {nextRedY, nextRedX}, {nextBlueY, nextBlueX}};
    } // end of proc

} // end of class

File: 백준/Gold/15644. 구슬 탈출 3/구슬 탈출 3.py
"""
4^10 = 2^20 = 10^6

완탐 가능

상하좌우 모든 경우의 수를 담은 배열을 만들고 돌려보면서 찾기

proc == 구슬 굴리기 함수
up이면 구슬부터 올라가서 위치를 찾으면 됨

그럼 어차피 보드에서 각 위치가 갈 수 있는 위치는 정해짐 그걸 찾아두자

틀림
N만큼 뻗어가면서 탐색하기로 했는데 만약 M이 더 길다면!!!!!
N과 M중에 더 긴걸로 하장
"""
import sys

input = lambda: sys.stdin.readline().strip()


def proc(d, red_position, blue_position): # 0: 갱신 완료 1: 성공 2: 실패
    red_y, red_x = red_position
    blue_y, blue_x = blue_position
    next_red_y, next_red_x = next_position[red_y][red_x][d]
    next_blue_y, next_blue_x = next_position[blue_y][blue_x][d]
    if next_blue_y == 20:
        return 2, [], []
    elif next_red_y == 20:
        return 1, [], []
    if next_red_y == next_blue_y and next_red_x == next_blue_x:
        if d == 0:
            if red_y > blue_y:
                next_red_y += 1
            else:
                next_blue_y += 1
        elif d == 1:
            if red_x > blue_x:
                next_blue_x -= 1
            else:
                next_red_x -= 1
        elif d == 2:
            if red_y > blue_y:
                next_blue_y -= 1
            else:
                next_red_y -= 1
        elif d == 3:
            if red_x > blue_x:
                next_red_x += 1
            else:
                next_blue_x += 1
    return 0, [next_red_y, next_red_x], [next_blue_y, next_blue_x]


def recur(cur, red_position, blue_position, history):
    global answer
    if cur == 10 or cur + 1 >= answer:
        return
    for i, d in [[0, 'U'], [1, 'R'], [2, 'D'], [3, 'L']]:
        result, next_red_position, next_blue_position = proc(i, red_position, blue_position)
        if result == 1:
            if answer > cur + 1:
                global answer_lst
                answer = cur + 1
                answer_lst = history + [d]
            return
        elif result == 2:
            continue
        recur(cur + 1, next_red_position, next_blue_position, history + [d])


N, M = map(int, input().split())
lst = []
for _ in range(N):
    lst.append(list(input()))

next_position = [[[[-1 for _ in range(2)] for _ in range(4)] for _ in range(M)] for _ in range(N)]
ori_red_y = -1
ori_red_x = -1
ori_blue_y = -1
ori_blue_x = -1
for i in range(1, N - 1):
    for j in range(1, M - 1):
        if lst[i][j] == 'R':
            ori_red_y = i
            ori_red_x = j
        elif lst[i][j] == 'B':
            ori_blue_y = i
            ori_blue_x = j
        # 위치 찾아두기
        for dy, dx, d in [[-1, 0, 0], [0, 1, 1], [1, 0, 2], [0, -1, 3]]: # 상 우 하 좌
            for k in range(max(N, M)):
                ny = i + dy * k
                nx = j + dx * k
                if 0 <= ny < N and 0 <= nx < M:
                    if lst[ny][nx] == 'O':
                        next_position[i][j][d] = [20, 20] # 골인인 경우
                        break
                    elif lst[ny][nx] != '#':
                        next_position[i][j][d] = [ny, nx]
                    else:
                        break

answer = 20
answer_lst = []
recur(0, [ori_red_y, ori_red_x], [ori_blue_y, ori_blue_x], [])

if answer != 20:
    print(answer)
    print(''.join(answer_lst))
else:
    print(-1)